Improve switch printing and row wire contact

BlueDrink9 opened the pull request:
Includes some refactor of switch shape to unify choc, mx and improved mx code.

Commit:
Make 5 pin holes optional

Commit:
DRY some switch code

Commit:
More DRY on wire channels

Commit:
Add teardrop module

Commit:
Use teardrops for wire channels

BlueDrink9 commented:
Quick pro tip: `git config merge.renormalize true` prevented a merge conflict here due to the line ending changes. Might be a good idea to suggest this to other PRs if they struggle with that change.

Commit:
Unify choc and mx socket cutouts

Commit:
Column wire crosses top pin at an angle to improve contact

Commit:
Move extra modules to utils

Commit:
Refactor kink code and fix distances

Commit:
Document new parameters

BlueDrink9 commented:
FYI, I'm planning on printing some tests of the standard MX switch holder some time in the next two weeks, so probably a good idea to not merge until I've indicated the changes work!

Commit:
Fix kink with row-staggered layout

50an6xy06r6n commented on scad/README.md:
```diff
@@ -37,7 +37,9 @@
 | `switch_orientation` | This can be set to either `north` or `south`, and defines the LED orientation of the switch. South-facing offers better compatibility with Cherry-profile keycaps, while north-facing may result in better illumination of shine-through legends (though there is not yet support for LEDs). | `parameters.scad` |
+| `five_pin_switch` | Sets whether to print the two side holes for the extra stability pins of a 5-pin switch. `true` fits 3 and 5 pin switches, `false` is cleaner and may be marginally faster to print. | `parameters.scad` |
 | `use_folded_contact` | This generates an experimental socket for MX switches that provides more robust hotswap contacts at the expense of more complexity when bending diodes. | `parameters.scad` |
+| `top_pin_wire_kink_angle` | This sets the angle that the row wire crosses over the top switch pin at, to improve contact between the sharp pin and the wires. Probably best not to go above 30 or below -35, unless you don't mind holes in walls. A larger angle may require cutting a larger hole in the wire casing. | `parameters.scad` |
```
```suggestion
| `row_wire_contact_angle` | This sets the angle that the row wire crosses over the top switch pin at, to improve contact between the sharp pin and the wires. Probably best not to go above 30 or below -35, unless you don't mind holes in walls. A larger angle may require cutting a larger hole in the wire casing. | `parameters.scad` |
```
What do you think of this naming? Feels more descriptive of its function

BlueDrink9 replied:
Yeah nice, heaps better

50an6xy06r6n commented on scad/parameters.scad:
```diff
@@ -144,7 +147,12 @@ h_unit = unit;
 v_unit = unit;
 // Spacing of grid for MX pins
-grid = 1.27;
+grid =
+    switch_type == "mx"
+    ? 1.27
+    : switch_type == "choc"
+        ? 0
+        : assert(false, "switch_type is invalid");
```
Is this needed if `grid` isn't used in the choc case anyway? I was thinking of just renaming this to `mx_schematic_unit` or something

BlueDrink9 replied:
Good point. Pretty sure I didn't make this variable(I think?), but happy to refactor while I'm there. Better double-check it isn't being added to anything without a mx check

50an6xy06r6n replied:
yeah I made this variable and was planning to rename it since `grid` is quite ambiguous. Then I saw your PR and wanted to make sure you weren't relying on it for anything.

BlueDrink9 replied:
I'll rename it when i adress these other things

50an6xy06r6n commented on scad/switch.scad:
```diff
@@ -24,185 +24,182 @@ module switch_socket_base(borders=[1,1,1,1]) {
-    }
-}
+    // Pin positions within socket cutout
+    top_pin_xy =
+        switch_type == "mx" ? [2*grid,4*grid]
+        : switch_type == "choc" ? [0,5.9]
+        : undef;
+    bottom_pin_xy =
+        switch_type == "mx" ? [-3*grid,2*grid]  // No mx bottom pin
+        : switch_type == "choc" ? [5,3.8]
+        : undef;
+    // Side pin is +-
+    side_pin_x =
+        switch_type == "mx" ? 4
+        : switch_type == "choc" ? 5.5
+        : undef;
+    diode_cutout_xy =
+        switch_type == "mx" ? [3*grid,-4*grid]
+        : switch_type == "choc" ? [-3.125, -3.8]
+        : undef;
+    col_channel_xy =
+        switch_type == "mx" ? [3*grid,-4*grid]
+        : switch_type == "choc" ? [-3.125,-3.8]
+        : undef;
+    row_channel_y =
+        switch_type == "mx" ? 4*grid
+        : switch_type == "choc" ? 5.9
+        : undef;
+    central_pin_r =
+        switch_type == "mx" ? 2.1
+        : switch_type == "choc" ? 1.75
+        : undef;
+    top_pin_cutout_r = 1;
```
I see what you're doing here, but I did like the modularity of having separate modules for the different switch types. What if we have a helper module that takes all these parameters explicitly and then wrap it in a module for each switch type that defines the values?

BlueDrink9 replied:
Fair enough, I had that then inlined them because it just added a bunch of shallow function calls with a lot of parameters. Happy to put them back. I think you're right, it'll end up a tidier way to define those values

50an6xy06r6n commented on scad/switch.scad:
```diff
@@ -24,185 +24,182 @@ module switch_socket_base(borders=[1,1,1,1]) {
+    }
+}
+
 module choc_socket_cutout(borders=[1,1,1,1], rotate_column=false) {
     render() translate([h_unit/2,-v_unit/2,0]) rotate([0,0,switch_rotation])
         intersection() {
             union() {
```
I assume you just forgot to clean this up?

50an6xy06r6n replied:
(note that I was proposing we bring it back lol)

50an6xy06r6n commented on scad/utils.scad:
```diff
@@ -1,5 +1,42 @@
 
+/* Creates a flat 2D teardrop on the XY plane,
+ * where no angle in the positive y direction is greater than 45º. */
+module teardrop2d(radius){
+	// Find tangents on the circle at 45 degrees
+	// Radius is triangle hypotenuse
+	function tangent_point(circle_r, angle) = [
+		circle_r * cos(angle),
+				 circle_r * sin(angle),
+	];
+	teardrop_point = [0,
+	tangent_point(radius, 45).y + tangent_point(radius, 45).x];
+	circle(radius);
+	polygon([
+		tangent_point(radius, 45),
+		tangent_point(radius, 135),
+		teardrop_point
+	]);
+}
```
Can you make the tangent angle configurable? I'm a little concerned about the tradeoff in wire retention

BlueDrink9 replied:
Absolutely. I personally struggled to get the wires in, so I wrecked my retention off the bat with print cleanup and my mental model of how retention should work is bad. Can you say a little more about how you picture retention working and how it interacts with the teardrop?

50an6xy06r6n replied:
In my head the retention is a function of how much the channel walls overhang, which I think in your code is `cos(angle)`. (Personally I think it makes sense to define the angle as deflection from vertical, so 0 degrees is a vertical wall). Assuming wire depth stays constant, the angle is the main parameter that determines the size of the channel opening, and that just seems like something that folks might want to configure, depending on how much trouble they have getting the wires to fit into or stay in the channels.

50an6xy06r6n replied:
In my head the retention is a function of how much the channel walls overhang, which I think in your code is `cos(angle)`. (Personally I think it makes sense to define the angle as deflection from vertical, so 0 degrees is a vertical wall). Assuming wire depth stays constant, the angle is the main parameter that determines the size of the channel opening, and that just seems like something that folks might want to configure, depending on how much trouble they have getting the wires to fit into or stay in the channels.

BlueDrink9 replied:
I think I kind of understand that. I'll make it customisable regardless, I think it makes sense even from the perspective that people tune printers to be able to do overhangs > 45º. 

Regarding retention (and I assume we're only talking about the uppermost channel here? (by default row)): This teardrop shape only gets wider than the equivalent cylinder once the tangent is at 45º, which I think is basically half way to a fully closed circle. I can't imagine anyone would have a wire falling out of something that narrow given the widest part of the circle is the wire diameter anyway. Does that sound right to you?

50an6xy06r6n commented on scad/parameters.scad:
```diff
@@ -16,9 +16,12 @@ pcb_type = "printed";  // [printed, traditional]
 use_folded_contact = false;
-
+// Angle the wire crosses the top switch pin
+top_pin_wire_kink_angle = 15;
```
Looking at the output, it feels like the default angle should be negative, since that avoids potential structural issues with the top of the channel. I see that it interferes a bit with the choc diode channel, but wrapping that channel the other way around the central pin should fix that.

BlueDrink9 replied:
Avoiding structural issues was why I suggested an upper bound of 35 in the docs, should probably have added that here too.

Negative hits the mx diode pin around 35°, positive hits the cutout edge around 40° so that's why I went with positive.

I have no other preference, but I'd be uncomfortable altering the choc shape. If you want to do that, would you be happy making the change yourself?

50an6xy06r6n replied:
yeah I can try to change the side that the choc diode wraps

BlueDrink9 replied:
Thanks. I don't have any choc switches with which to test this

BlueDrink9 replied:
I'll leave the choc diode change to you in a separate PR I think.

50an6xy06r6n commented on scad/switch.scad:
```diff
@@ -24,185 +24,182 @@ module switch_socket_base(borders=[1,1,1,1]) {
-                        cylinder(h=pcb_thickness+1,r=1.05);
+                if (five_pin_switch){
+                    for (x = [-side_pin_x, side_pin_x]) {
+                        translate([x*grid,0,pcb_thickness/2-socket_depth])
```
the use of `grid` here causes the holes to not generate for chocs. I guess this makes more sense if you make the grid size `1` for chocs instead of `0`.

BlueDrink9 replied:
Oops, missed this in testing. Thanks

50an6xy06r6n replied:
actually the more I think about it, the more I think `side_pin_x` should always be in `mm` to match the rest of the parameters, so this module won't depend on `grid` at all

BlueDrink9 replied:
Should be able to work that myself

BlueDrink9 replied:
The temporary fix here anyway would be to move the `*grid` to where the side pin x for mx is defined (i.e. define it as `4*grid`

50an6xy06r6n reviewed (changes requested):
Looks great overall! Just had some thoughts on naming/code structure/defaults and found a bug with choc sockets.

BlueDrink9 commented:
Great review, thanks! I expect to have time to address these towards the end of the week.
I noted one thing I think it might be better for you to address if you have the capacity to

Commit:
Fix missing top switch pin hole

Commit:
Boost kink width to improve angle

Commit:
Rename row_wire_contact_angle

Commit:
Fix choc side pins not generating

Commit:
Allow customising teardrop overhang angle

Commit:
Change row angle default to -ve

Commit:
Rename mx_schematic_unit, refactor socket_cutouts

BlueDrink9 commented:
rebased against the latest master and addressed your requests. I believe this should now be good to merge.

Commit:
Tweak build guide based on my experiences.

Removed warning about overhang issues, because they should be resolved
by the teardrop

Commit:
Advise against crossing diode wires

Realised this when making my board, when I had ROW2COL/COL2ROW the wrong way
around in firmware, and half my keys didn't work (or so I thought -
actually none of them should have). The keys that worked were being
shorted by the crossed diodes.
This is a subtle mistake, because you won't notice it unless you're a
furious typer, but it defeats the point of having diodes and can lead to
ghosting when multiple keys are pressed at the same time.

BlueDrink9 commented:
Just updated the build guide, too. Take a look.

Second build is well under way, I'll be certain to post my experiences. Dactyl format makes using this a lot more work, second-guessing whether this will be worth the hotswap capability

BlueDrink9 commented:
Wired my board. This PR is good, think it's ready to merge @50an6xy06r6n ?

50an6xy06r6n reviewed (approved):
Thanks for your work on this! Looks good to me. Sorry I've been MIA for a while
